File: example/cpp/DemoOverlay.cpp
/**
 * @file DemoOverlay.cpp
 * @author BambooSlips
 * @brief 
 * @version 0.1
 * @date 2024-09-08
 * 
 * @copyright Copyright (c) 2024
 * 
 */

#include "common_def.hpp"
#include "DemoOverlay.hpp"
#include "MirageLog.hpp"
#include "MirageFont.hpp"

DemoOverlay::DemoOverlay()
           : BaseLayer("DemoOverlay")
           , m_Observer {}
           , m_UIImage {}
           , m_UIText {}
           , m_UIButton {}
{

}

DemoOverlay::DemoOverlay(std::string_view svDataDir)
           : DemoOverlay()
{
    m_strDataDir = svDataDir;

    Mirage::vec3<float> vCamFocusVec { 0, 0, 0 };
    Mirage::vec3<float> vCamTopVec   { 0, 1, 0 };
    Mirage::vec3<float> vCamTransPos { 0, 0, -1 };

    m_Observer.SetDepth(1);
    m_Observer.SetCullingMask(0x02);
    m_Observer.SetClearFlag(MRG_DEPTH_BUFFER_BIT);
    m_Observer.EnableOrthoView(-400.0f, 400.0f, 300.0f, -300.0f, 0.01f, 100.0f);
    m_Observer.SetPosition(vCamTransPos);
    m_Observer.SetView(vCamFocusVec, vCamTopVec);

    CreateUIText();
    CreateUIImage();
    CreateUIButton();
}

void DemoOverlay::OnAttach()
{

}

void DemoOverlay::OnDetach()
{

}

void DemoOverlay::OnUpdate()
{
    auto fnRender = [&]() { 
        m_Observer.ClearView();
        m_UIImage.Refresh();
        m_UIText.Refresh();
        m_UIButton.Refresh();
    };
    fnRender();
}

void DemoOverlay::OnEvent(Mirage::BaseEvent& rEvent)
{

}

void DemoOverlay::CreateUIText()
{
    unsigned char uLayer        { 0x2 };
    std::string strFontPath     { "font/OldEnglish/OLDENGL.TTF" };
    std::string strText         { "UI Text" };
    // TODO: remove the hard code
    float fViewWidth            { 800 };
    float fViewHeight           { 600 };
    float x                     { 0 };
    float y                     { 0 };
    Mirage::vec3<float> vTxtPos { 0, 0, 0 };
    Mirage::vec4<float> vTxtColor { 255, 215, 0, 1 };   // RGBA

    m_UIText.SetFontSize(60);
    // m_UIText.Configure(uLayer, strFontPath, strText);
    m_UIText.Configure(uLayer, strFontPath, strText, Mirage::TextAlign::left);

    x       = m_UIText.GetPitch() / 2;
    y       = fViewHeight / 2.0f;
    vTxtPos = { x, y, 0 };
    m_UIText.SetPosition(vTxtPos, true);
    m_UIText.SetColor(vTxtColor);
}

void DemoOverlay::CreateUIImage()
{
    int                 nHeight     { 100 };
    int                 nWidth      { 100 };
    std::string         strMaterial { "material/flower.mat" };
    unsigned char       uLayer      { 0x2 };
    // // TODO: remove the hard code
    float               fViewWidth  { 800 };
    float               fViewHeight { 600 };
    float               x           { fViewWidth - nWidth / 2 };
    float               y           { fViewHeight / 2 };
    Mirage::vec3<float> vScale      { 1.0 , 1.0, 1.0 };
    Mirage::vec3<float> vImgPos     { x, y, 0.0f };

    m_UIImage.Configure(nWidth, nHeight, uLayer, strMaterial);
    // m_UIImage.Configure(nWidth, nHeight, uLayer);
    m_UIImage.SetScale(vScale);
    m_UIImage.SetPosition(vImgPos, true);
    m_UIImage.AddMask("material/circle_mask.mat");
}

void DemoOverlay::CreateUIButton()
{
    int                 nWidth      { 40 };
    int                 nHeight     { 30 };
    unsigned char       uLayer      { 0x2 };
    // // TODO: remove the hard code
    float               fViewWidth  { 800 };
    float               fViewHeight { 600 };
    // Mirage::vec4<float> vTxtColor   { 255, 255, 255, 1 };   // RGBA
    float               x           { fViewWidth - nWidth / 2 };
    float               y           { fViewHeight / 2 - 100 };
    Mirage::vec3<float> vScale      { 1.0 , 1.0, 1.0 };
    Mirage::vec3<float> vButtonPos  { x, y, 0.0f };

    m_UIButton.Configure(nWidth, nHeight, uLayer);
    m_UIButton.SetScale(vScale);
    m_UIButton.SetPosition(vButtonPos, true);
    // m_UIButton.SetTextColor(vTxtColor);
    m_UIButton.SetClickCallback([&]() {
        bool bEnableMask = m_UIImage.IsMaskEnabled();
        m_UIImage.SetMaskStatus(!bEnableMask);
    });
}

File: example/python/DemoOverlay.py
import sys
from pathlib import Path

mirage_path = Path(__file__).parent.parent.parent / "bindings" / "python" / "src" / "Mirage"
sys.path.insert(0, str(mirage_path))
print("module example is in: ", mirage_path)

import Mirage

class DemoOverlay(Mirage.BaseLayer):
    def __init__(self):
        super().__init__()
        self.__observer = Mirage.Observer()
        self.__image    = Mirage.Image()
        self.__text     = Mirage.Text()
        self.__button   = Mirage.Button()
        self.__InitObserver()
        self.__CreateUIImg()
        self.__CreateUIText()
        self.__CreateUIButton()
    
    def OnAttach(self):
        print("OnAttach")
    
    def OnDetach(self):
        print("OnDetach")
    
    def OnUpdate(self):
        self.__observer.ClearView()
        self.__image.Refresh()
        self.__text.Refresh()
        self.__button.Refresh()
    
    def OnEvent(self, refEvent):
        print("OnEvent")
    
    def __InitObserver(self):
        position = Mirage.vec3_float(0, 0, -1)
        focusVec = Mirage.vec3_float(0, 0, 0)
        topVec   = Mirage.vec3_float(0, 1, 0)
        self.__observer.SetDepth(1)
        self.__observer.SetCullingMask(0x02)
        self.__observer.SetClearFlag(Mirage.MRG_DEPTH_BUFFER_BIT)
        self.__observer.EnableOrthoView(-400, 400, 300, -300, 0.01, 100)
        self.__observer.SetPosition(position)
        self.__observer.SetView(focusVec, topVec)
    
    def __CreateUIText(self):
        layer      = 0x02
        fontPath   = "font/OldEnglish/OLDENGL.TTF"
        text       = "UI Text"
        viewWidth  = 800
        viewHeight = 600

        self.__text.SetFontSize(60)
        self.__text.Configure(layer, fontPath, text, Mirage.left)

        x = self.__text.GetPitch() / 2
        y = viewHeight / 2.0

        position = Mirage.vec3_float(x, y , 0)
        color    = Mirage.vec4_float(255, 215, 0, 1)
        self.__text.SetPosition(position, True)
        self.__text.SetColor(color)
    
    def __CreateUIImg(self):
        height        = 100
        width         = 100
        material      = "material/flower.mat"
        mask_material = "material/circle_mask.mat"
        layer         = 0x02
        viewWidth     = 800
        viewHeight    = 600
        scale         = Mirage.vec3_float(1.0, 1.0, 1.0)
        x             = viewWidth - width / 2
        y             = viewHeight / 2
        position      = Mirage.vec3_float(x, y, 0.0)

        self.__image.Configure(width, height, layer, material)
        self.__image.AddMask(mask_material)
        self.__image.SetScale(scale)
        self.__image.SetPosition(position, True)
    
    def __CreateUIButton(self):
        width      = 40
        height     = 30
        layer      = 0x02
        viewWidth  = 800
        viewHeight = 600
        x          = viewWidth - width / 2
        y          = viewHeight / 2 - 100
        scale      = Mirage.vec3_float(1.0, 1.0, 1.0)
        position   = Mirage.vec3_float(x, y, 0.0)

        self.__button.Configure(width, height, layer)
        self.__button.SetScale(scale)
        self.__button.SetPosition(position, True)
        self.__button.SetClickCallback(
            lambda : self.__image.SetMaskStatus(not self.__image.IsMaskEnabled())
        )